### PR TITLE
Bug 1836778 - Add Support for Printing in GeckoEngineSession

### PR DESCRIPTION
**Pull request by ohall-m**
This bug adds an option to print the current page's content using the Android Print Spooler via GeckoView with status on the Engine in AC.

GeckoView bug [1837426](https://bugzilla.mozilla.org/show_bug.cgi?id=1837426) adds in the new `didPrintPageContent` to the GeckoView API first.

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1836778

### CODE REVIEW

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 1]
```diff
@@ -276,6 +276,29 @@ class GeckoEngineSession(
         )
     }
 
+    /**
+     * See [EngineSession.requestPrintContent]
+     */
+    override fun requestPrintContent() {
+        geckoSession.didPrintPageContent().then(
```
Comment (jonalmeida) [inline android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:283]: Note aside: at some point, consider if it's valuable to move these actions to a `PrintController` so we have all the print actions on that if we are going to add more in-browser printing APIs.

```kotlin
geckoSession.printController.saveAsPdf()
geckoSession.printController.didPrintPageContent()
geckoSession.printController.printPageContent()
// etc..
```

**Review by jonalmeida — APPROVED** [review]
LGTM with some minor nits!

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:283]: I think this is a good idea, I have [this](https://github.com/mozilla-mobile/firefox-android/pull/2317) open right now for more context too. Now that you mention it, both use-cases are converging into the same pattern.

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 2]
```diff
@@ -283,6 +283,18 @@ abstract class EngineSession(
...
+         * Event to indicate that printing finished.
+         */
+        fun onPrintFinish() = Unit
+
+        /**
+         * Event to indicate that an exception was thrown while preparing to print.
+         *
+         * @param throwable The exception throwable. Usually a GeckoPrintException.
+         */
+        fun onPrintException(throwable: Throwable) = Unit
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt:296]: I like that idea, they are similar enough and essentially have the same error conditions. I'll rename `onSaveToPdfException` to `onGeckoPrintException`, and add a Print/PDF flag once #2317 is submitted.

**Comment by ohall-m** [comment]
Will rebase for landing after bug #2317

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt [hunk 3]
```diff
@@ -276,6 +276,29 @@ class GeckoEngineSession(
...
+            { finishedPrinting ->
+                if (finishedPrinting == true) {
+                    notifyObservers {
+                        onPrintFinish()
+                    }
+                }
+                GeckoResult<Boolean>()
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt]: Thanks, updated!

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: Thanks, I renamed this one to `onGeckoPrintException(isPrint: Boolean, throwable: Throwable)`. When I implement the UI for printing, I'll remove the `PdfException` to become integrated into `onGeckoPrintException(isPrint:false)`. 

What do you think?

File: android-components/components/feature/session/src/main/java/mozilla/components/feature/session/SessionUseCases.kt [hunk 4]
```diff
@@ -432,6 +432,35 @@ class SessionUseCases(
...
+    /**
+     * A use case for requesting a given tab to print it's content.
+     */
+    class PrintContentUseCase internal constructor(
+        private val store: BrowserStore,
+    ) {
+        /**
+         * Request that Android print the current [tabId].
+         *
+         * The same cavets in the [SaveToPdfUseCase] apply here because GeckoView makes a PDF prior
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/feature/session/src/main/java/mozilla/components/feature/session/SessionUseCases.kt]: Thanks!

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:283]: I like this plan, I think a good opportunity to refactor into a controller will be [bug 1838437](https://bugzilla.mozilla.org/show_bug.cgi?id=1838437) or when I implement the print UI.

**Review by jonalmeida — APPROVED** [review]
Still LGTM!

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonalmeida) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: Started a new thread in error.

Yeah, this sounds fine!

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt [hunk 5]
```diff
@@ -283,6 +283,19 @@ abstract class EngineSession(
...
+         */
+        fun onPrintFinish() = Unit
+
+        /**
+         * Event to indicate that an exception was thrown while preparing to print.
+         *
+         * @param isPrint true for a true print error or false for a Save as PDF error.
+         * @param throwable The exception throwable. Usually a GeckoPrintException.
+         */
+        fun onGeckoPrintException(isPrint: Boolean, throwable: Throwable) = Unit
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/EngineSession.kt]: Thanks, adjusted!

File: android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt (hunk 1 above)
Comment (ohall-m) [reply android-components/components/browser/engine-gecko/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt:284]: Filed [bug 1840715](https://bugzilla.mozilla.org/show_bug.cgi?id=1840715)! (Print UI patch was getting kinda big for this refactor to be included).

**Pull request merged**